Can label features with two colors

msekatchev opened the pull request:
Features can be labeled with blue or red.
Mask is saved with two colors, (1,1,1) for featured drawn in blue and (2,2,2) for features drawn in red.
<del>Color is toggled by pressing C.</del>

msekatchev committed:
Can label features with two colors

Features can be labeled with blue or red.
Color is toggled by pressing C.

pdeperio commented on annotate.py:
```diff
@@ -102,7 +115,13 @@ def annotate_img(img_path, initials, size1=1, size2=1) :
             break
 
-        
+        if(k==ord('c')):
```
Why provide this (redundant) functionality, if left vs middle click already does this?

msekatchev replied:
Yes I agree it's redundant, I'll simplify the code and use left/middle click instead

msekatchev replied:
Fixed in new version of PR

pdeperio commented on annotate.py:
```diff
@@ -188,20 +207,36 @@ def annotate_img(img_path, initials, size1=1, size2=1) :
-    mask = cv2.inRange(train_labels, lower, upper)
-    mask[mask < 250] = 0
-    mask[mask != 0 ] = 255 ##set =1 for binary or =255 for visible white
+    lower_blue = np.array([254,0,0], dtype = "uint16")
```
These numbers are hardcoded and repeated above in `draw_circle` (i.e. need two places to modify if changing, which can lead to mistakes). Find a way to use a single variable here and above (probably `global` is fine for this small script, though generally bad practice).

msekatchev replied:
Added global variables in the new version of PR

pdeperio commented:
please also comment if this is all the functionality from:

https://github.com/martindgr/make_masks/blob/master/draw_mask.py

such that @martindgr can start using this script instead.

is his directory structure not needed anymore, i.e. can use the now simplified one since we won't be dealing with crops anymore?

msekatchev committed:
Can label features with two colors

Features can be labeled with blue (middle-click) or red (left-click).

msekatchev commented:
> please also comment if this is all the functionality from:
> 
> https://github.com/martindgr/make_masks/blob/master/draw_mask.py
> 
> such that @martindgr can start using this script instead.
> 
> is his directory structure not needed anymore, i.e. can use the now simplified one since we won't be dealing with crops anymore?

Yes, this is all the same functionality, though the buttons have been changed:
 - middle-click to draw PMTs (blue) and left-click to draw PMT bolts (red).

I would suggest moving to the simplified directory structure that we have in this version.

pdeperio commented on annotate.py:
```diff
@@ -212,20 +214,19 @@ def annotate_img(img_path, initials, size1=1, size2=1) :
-    mask_blue[mask_blue < 250] = 0
-    #mask_blue[mask_blue != 0 ] = 255
-    mask_blue[mask_blue != 0 ] = 1
+    lower_red = np.array(RED_MIN, dtype = "uint16")
```
Since this is just setting a range for finding the pixels that you've already precisely defined with `RED_MAX`, I guess you don't need this extra `RED_MIN` variable, and can just subtract 1 from `RED_MAX` like:
```
tuple(np.subtract(RED_MAX, np.divide(RED_MAX, RED_MAX, out=np.zeros_like(RED_MAX), where=RED_MAX!=0)))
```
(sorry complicated tuple algebra; make sure to add a comment in the code what this is doing!  maybe you can find a nicer way, but should be good for generality with `BLUE_MAX` too below)

msekatchev replied:
The code seems to work if both the upper and lower limits are equal (both `255` instead of a range).

We could just have:
`upper_red = np.array(RED_MAX, dtype = "uint16")`
    
`mask_red = cv2.inRange(train_labels, upper_red, upper_red)`

Do we still need to find `RED_MIN`?

pdeperio replied:
nope, even better.

msekatchev replied:
Ok nice that simplifies things.
(I'm using this method in `find_centroids.py` too)

pdeperio commented on annotate.py:
```diff
@@ -212,20 +214,19 @@ def annotate_img(img_path, initials, size1=1, size2=1) :
+    lower_red = np.array(RED_MIN, dtype = "uint16")
+    upper_red = np.array(RED_MAX, dtype = "uint16")
+    mask_red = cv2.inRange(train_labels, lower_red, upper_red)
+    mask_red[mask_red < COLOR_THRESHOLD] = 0
```
Similarly for `COLOR_THRESHOLD`, I'm not sure in what case you will have a pixel between `250` and `254`, so why not just use:
```
RED_MAX[np.nonzero(RED_MAX)[0]]
```

msekatchev replied:
Same here, seems to work with `COLOR_THRESHOLD = 255`.

pdeperio replied:
great, then you can get rid of the `COLOR_THRESHOLD` variable too (see next comment as well).

pdeperio commented on annotate.py:
```diff
@@ -212,20 +214,19 @@ def annotate_img(img_path, initials, size1=1, size2=1) :
-    upper_red = np.array([0,0,255], dtype = "uint16")
-    mask_red = cv2.inRange(train_labels, lower_red, upper_red)
-    mask_red[mask_red < 250] = 0
+    lower_blue = np.array(BLUE_MIN, dtype = "uint16")
```
From here, this is actually more copy/pasted code from the `RED` section. Considering the above comments, in principle the only input you really need is `255`, which can then be used to define the rest of the variables, including both colours.

Consider using a list or dictionary that you can loop through, using only a single block of code. This will also help for generality, in case later we want to add a third feature in the `green` channel (or beyond with any number of colours, but will need to update the `np` tricks above).

msekatchev replied:
Sounds good, I'll change the code to have it loop through a dictionary of colors, and only keep one `255` variable that affects the rest.

msekatchev committed:
Can label features with two colors

Features can be labeled with blue (middle-click) or red (left-click).
Changed the way the mask image is saved to make it easier to add more colors later on.

pdeperio commented on README.txt:
```diff
@@ -53,6 +61,8 @@ def draw_circle(event,x,y,flags,param):
         ## s to close image and exit program
         ## r to write coordinates to file (you will be prompted for the PMT feature ID)
+        ## f to select new PMT
+        ## c to switch labelling color between blue and red.
```
delete remnant

pdeperio commented on README.txt:
```diff
@@ -3,38 +3,46 @@ import os
 
+#annotate_img("B.jpg","MS",10,120)
+
+global COLOR
```
Mods in `README.txt` by accident again?

msekatchev replied:
Sorry about that, I updated it.

msekatchev replied:
Updated README with Markdown formatting.

msekatchev committed:
Temporary readme

Working on proper readme in Markdown format.

msekatchev committed:
Can label features with two colors

Features can be labeled with blue (middle-click) or red (left-click).
Changed the way the mask image is saved to make it easier to add more colors later on.

msekatchev committed:
Updated README with Markdown formatting

msekatchev committed:
Delete README.txt

pdeperio commented:
@martindgr please review this PR, checkout and test. it should contain all your functionality for making masks, together with ID labeling. Folder structure has been simplified (so it's not too rigidly defined in the code, but just based on user's single input directory).

`find_centroids.py` and auto-labeling are forthcoming in a separate PR.

pdeperio commented:
@msekatchev please check [the README.md preview](https://github.com/WatChPhoto/pmt-annotation/blob/078184f51383336ff3a245276faa824d85856153/README.md), seems the headers aren't working. might need a space after `#`?

martindgr reviewed (commented):
Looking good, directory system is fine, as slicing and dataset making for neural net is going to be done after this I expect.
 The only problem I have is the loss of the ability to click and drag to paint which is pretty vital for making accurate masks (notable getting clean edges around the pmt is very hard to do with a single click and tedious with many clicks). 
Is there a way to reimplement this without breaking it?  I'm guessing for it to work the recording of mouse position will have to be turned off since it returns the mouse position at every step the mouse is held down, maybe on a shift+click+drag? Or just record the mouse click on the initial LBUTTONDOWN and not on MOUSEMOVE?

msekatchev committed:
Update README.md

msekatchev commented:
> @msekatchev please check [the README.md preview](https://github.com/WatChPhoto/pmt-annotation/blob/078184f51383336ff3a245276faa824d85856153/README.md), seems the headers aren't working. might need a space after `#`?

Yeah it's fixed, I updated it on my branch but forgot to do it here

msekatchev commented:
> Looking good, directory system is fine, as slicing and dataset making for neural net is going to be done after this I expect.
> The only problem I have is the loss of the ability to click and drag to paint which is pretty vital for making accurate masks (notable getting clean edges around the pmt is very hard to do with a single click and tedious with many clicks).
> Is there a way to reimplement this without breaking it? I'm guessing for it to work the recording of mouse position will have to be turned off since it returns the mouse position at every step the mouse is held down, maybe on a shift+click+drag? Or just record the mouse click on the initial LBUTTONDOWN and not on MOUSEMOVE?

Ok I can work on reimplementing this

pdeperio commented on annotate.py:
```diff
@@ -47,6 +61,8 @@ def draw_circle(event,x,y,flags,param):
         ## click on image to grab pixel coordinates
         ## s to close image and exit program
         ## r to write coordinates to file (you will be prompted for the PMT feature ID)
+        ## f to select new PMT
```
update with all keystroke features (or just remove to only need to maintain the README)

pdeperio commented on README.md:
```diff
@@ -0,0 +1,52 @@
+* click on image to grab pixel coordinates
+* s to close image and exit program
+* r to write coordinates to file (you will be prompted for the PMT feature ID)
+* f to select new PMT
```
update with all new keystroke features

pdeperio commented:
@msekatchev reminder to address all comments before merging and let someone else merge your code ;)

msekatchev commented:
> @msekatchev reminder to address all comments before merging and let someone else merge your code ;)

Ok sorry about that, I'll make a new PR addressing user instructions.

martindgr commented:
I have had a little play around and it seems that just dropping in the below lines into the callbacks (similar to in my original) works and only records the first click of the mouse to save to .txt but you may want to text for yourselves.

`        elif event == cv2.EVENT_MOUSEMOVE:
            if drawing == True:
                cv2.circle(img, (x, y), small_size,BLUE,-1)`
 (similar for middle mouse)

just a small thing but the labels for RED/BLUE  are blue/red respectively since cv2 uses bgr instead of rgb

msekatchev commented:
> I have had a little play around and it seems that just dropping in the below lines into the callbacks (similar to in my original) works and only records the first click of the mouse to save to .txt but you may want to text for yourselves.
> 
> ` elif event == cv2.EVENT_MOUSEMOVE: if drawing == True: cv2.circle(img, (x, y), small_size,BLUE,-1)`
> (similar for middle mouse)
> 
> just a small thing but the labels for RED/BLUE are blue/red respectively since cv2 uses bgr instead of rgb

Yes this works, thank you Dan. I also changed the RED/BLUE labels and will make a new PR once we merge the updated user instruction changes.
